File: android/src/main/java/com/mattermost/networkclient/helpers/DocumentHelper.java
package com.mattermost.networkclient.helpers;

import com.mattermost.networkclient.APIClientModule;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.os.ParcelFileDescriptor;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.provider.OpenableColumns;
import android.text.TextUtils;
import android.webkit.MimeTypeMap;


import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;

public class DocumentHelper {
    private final static Context context = APIClientModule.context;
    private final static String CACHE_DIR_NAME = "NetworkClientCacheDir";

    public static String getRealPath(final Uri uri) {
        if (DocumentsContract.isDocumentUri(context, uri)) {
            if (isExternalStorageDocument(uri)) {
                final String docId = DocumentsContract.getDocumentId(uri);
                final String[] split = docId.split(":");
                final String type = split[0];

                if ("primary".equalsIgnoreCase(type)) {
                    return Environment.getExternalStorageDirectory() + "/" + split[1];
                }
            } else if (isDownloadsDocument(uri)) {
                final String id = DocumentsContract.getDocumentId(uri);
                if (!TextUtils.isEmpty(id)) {
                    if (id.startsWith("raw:")) {
                        return id.replaceFirst("raw:", "");
                    }
                    try {
                        return getPathFromSavingTempFile(context, uri);
                    } catch (NumberFormatException e) {
                        return null;
                    }
                }
            } else if (isMediaDocument(uri)) {
                final String docId = DocumentsContract.getDocumentId(uri);
                final String[] split = docId.split(":");
                final String type = split[0];

                Uri contentUri = null;
                if ("image".equals(type)) {
                    contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
                } else if ("video".equals(type)) {
                    contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
                } else if ("audio".equals(type)) {
                    contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
                }

                final String selection = "_id=?";
                final String[] selectionArgs = new String[] {
                        split[1]
                };

                if (contentUri != null) {
                    return getDataColumn(context, contentUri, selection, selectionArgs);
                } else {
                    return getPathFromSavingTempFile(context, uri);
                }
            }
        }

        if ("content".equalsIgnoreCase(uri.getScheme())) {
            if (isGooglePhotosUri(uri)) {
                return uri.getLastPathSegment();
            }

            return getPathFromSavingTempFile(context, uri);
        } else if ("file".equalsIgnoreCase(uri.getScheme())) {
            return uri.getPath();
        }

        return null;
    }

    public static String getPathFromSavingTempFile(Context context, final Uri uri) {
        File tmpFile;
        String fileName = null;

        if (uri == null || uri.isRelative()) {
            return null;
        }

        try {
            Cursor returnCursor =
                    context.getContentResolver().query(uri, null, null, null, null);
            int nameIndex = returnCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
            returnCursor.moveToFirst();
            fileName = sanitizeFilename(returnCursor.getString(nameIndex));

        } catch (Exception e) {
            // just continue to get the filename with the last segment of the path
        }

        try {
            if (TextUtils.isEmpty(fileName)) {
                fileName = sanitizeFilename(uri.getLastPathSegment().toString().trim());
            }


            File cacheDir = new File(context.getCacheDir(), CACHE_DIR_NAME);
            if (!cacheDir.exists()) {
                cacheDir.mkdirs();
            }

            String mimeType = getMimeType(uri.getPath());
            tmpFile = new File(cacheDir, fileName);
            tmpFile.createNewFile();

            ParcelFileDescriptor pfd = context.getContentResolver().openFileDescriptor(uri, "r");

            FileChannel src = new FileInputStream(pfd.getFileDescriptor()).getChannel();
            FileChannel dst = new FileOutputStream(tmpFile).getChannel();
            dst.transferFrom(src, 0, src.size());
            src.close();
            dst.close();
        } catch (IOException ex) {
            return null;
        }
        return tmpFile.getAbsolutePath();
    }

    public static String getDataColumn(Context context, Uri uri, String selection,
                                       String[] selectionArgs) {
        Cursor cursor = null;
        final String column = "_data";
        final String[] projection = {
                column
        };

        try {
            cursor = context.getContentResolver().query(uri, projection, selection, selectionArgs,
                    null);
            if (cursor != null && cursor.moveToFirst()) {
                final int index = cursor.getColumnIndexOrThrow(column);
                return cursor.getString(index);
            }
        } finally {
            if (cursor != null)
                cursor.close();
        }
        return null;
    }


    public static boolean isExternalStorageDocument(Uri uri) {
        return "com.android.externalstorage.documents".equals(uri.getAuthority());
    }

    public static boolean isDownloadsDocument(Uri uri) {
        return "com.android.providers.downloads.documents".equals(uri.getAuthority());
    }

    public static boolean isMediaDocument(Uri uri) {
        return "com.android.providers.media.documents".equals(uri.getAuthority());
    }

    public static boolean isGooglePhotosUri(Uri uri) {
        return "com.google.android.apps.photos.content".equals(uri.getAuthority());
    }

    public static String getExtension(String uri) {
        if (uri == null) {
            return null;
        }

        int dot = uri.lastIndexOf(".");
        if (dot >= 0) {
            return uri.substring(dot);
        } else {
            // No extension.
            return "";
        }
    }

    public static String getMimeType(File file) {

        String extension = getExtension(file.getName());

        if (extension.length() > 0)
            return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension.substring(1));

        return "application/octet-stream";
    }

    public static String getMimeType(String filePath) {
        File file = new File(filePath);
        return getMimeType(file);
    }

    private static String sanitizeFilename(String filename) {
        if (filename == null) {
            return null;
        }

        File f = new File(filename);
        return f.getName();
    }
}

File: android/src/main/java/com/mattermost/networkclient/helpers/DocumentHelper.kt
package com.mattermost.networkclient.helpers

import android.content.Context
import android.net.Uri
import android.os.Environment
import android.provider.DocumentsContract
import android.provider.MediaStore
import android.provider.OpenableColumns
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

class DocumentHelper private constructor() {
    companion object {
        private const val CACHE_DIR_NAME = "NetworkClientCacheDir"
        @Volatile
        private var instance: DocumentHelper? = null

        fun getInstance(): DocumentHelper {
            return instance ?: synchronized(this) {
                instance ?: DocumentHelper().also { instance = it }
            }
        }
    }

    fun getRealPath(context: Context, uri: Uri): String? {
        if (DocumentsContract.isDocumentUri(context, uri)) {
            when {
                isExternalStorageDocument(uri) -> {
                    val docId = DocumentsContract.getDocumentId(uri)
                    val split = docId.split(":").toTypedArray()
                    val type = split[0]

                    if ("primary".equals(type, ignoreCase = true)) {
                        return "${Environment.getExternalStorageDirectory()}/${split[1]}"
                    }
                }
                isDownloadsDocument(uri) -> {
                    val id = DocumentsContract.getDocumentId(uri)
                    if (!id.isNullOrEmpty()) {
                        return if (id.startsWith("raw:")) {
                            id.removePrefix("raw:")
                        } else {
                            try {
                                return getPathFromSavingTempFile(context, uri)
                            } catch (e: NumberFormatException) {
                                null
                            }
                        }
                    }
                }
                isMediaDocument(uri) -> {
                    val docId = DocumentsContract.getDocumentId(uri)
                    val split = docId.split(":").toTypedArray()
                    val type = split[0]

                    val contentUri = when (type) {
                        "image" -> MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                        "video" -> MediaStore.Video.Media.EXTERNAL_CONTENT_URI
                        "audio" -> MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
                        else -> null
                    }

                    val selectionArgs = arrayOf(split[1])

                    return if (contentUri != null) {
                        getDataColumn(context, contentUri, selectionArgs)
                    } else {
                        getPathFromSavingTempFile(context, uri)
                    }
                }
            }
        }

        when {
            "content".equals(uri.scheme, ignoreCase = true) -> {
                return if (isGooglePhotosUri(uri)) {
                    uri.lastPathSegment
                } else {
                    getPathFromSavingTempFile(context, uri)
                }
            }
            "file".equals(uri.scheme, ignoreCase = true) -> {
                return uri.path
            }
        }

        return null
    }

    private fun getPathFromSavingTempFile(context: Context, uri: Uri): String? {
        var tmpFile: File? = null
        var fileName: String? = null

        if (uri.isRelative) {
            return null
        }

        try {
            val returnCursor = context.contentResolver.query(uri, null, null, null, null)
            returnCursor?.use {
                val nameIndex = it.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                it.moveToFirst()
                fileName = sanitizeFilename(it.getString(nameIndex))
            }
        } catch (e: Exception) {
            // just continue to get the filename with the last segment of the path
        }

        try {
            if (fileName.isNullOrEmpty()) {
                fileName = sanitizeFilename(uri.lastPathSegment.toString().trim())
            }

            val cacheDir = File(context.cacheDir, CACHE_DIR_NAME)
            if (!cacheDir.exists()) {
                cacheDir.mkdirs()
            }

            if (!fileName.isNullOrEmpty()) {
                tmpFile = File(cacheDir, fileName!!)
                tmpFile.createNewFile()

                val pfd = context.contentResolver.openFileDescriptor(uri, "r")
                FileInputStream(pfd!!.fileDescriptor).channel.use { src ->
                    FileOutputStream(tmpFile).channel.use { dst ->
                        dst.transferFrom(src, 0, src.size())
                    }
                }
                pfd.close()
            }
        } catch (ex: IOException) {
            return null
        }

        return tmpFile?.absolutePath
    }

    private fun getDataColumn(context: Context, uri: Uri, selectionArgs: Array<String>?): String? {
        val selection = "_id=?"
        val column = "_data"
        val projection = arrayOf(column)

        context.contentResolver.query(uri, projection, selection, selectionArgs, null)?.use { cursor ->
            if (cursor.moveToFirst()) {
                val index = cursor.getColumnIndexOrThrow(column)
                return cursor.getString(index)
            }
        }
        return null
    }

    private fun isExternalStorageDocument(uri: Uri): Boolean {
        return "com.android.externalstorage.documents" == uri.authority
    }

    private fun isDownloadsDocument(uri: Uri): Boolean {
        return "com.android.providers.downloads.documents" == uri.authority
    }

    private fun isMediaDocument(uri: Uri): Boolean {
        return "com.android.providers.media.documents" == uri.authority
    }

    private fun isGooglePhotosUri(uri: Uri): Boolean {
        return "com.google.android.apps.photos.content" == uri.authority
    }

    private fun sanitizeFilename(filename: String?): String? {
        filename ?: return null
        val file = File(filename)
        return file.name
    }
}

